### ISSUE TITLE
Unused arguments to function changes output

### ISSUE THREAD

**Issue by chrisc36** [google/jax#12714]
### Description

After a long time tracking down a NaN in a neural network training on a TPU, I was able to isolate a case where the output of the jit-compiled forward/backward pass of the network changes depending on whether/what extra arrays, that are never used in the actual computations, are passed into the function as additional arguments and then returned unchanged. Depending on exactly what unused arrays are passed in I have seen:

- Sometimes the mean gradient will become NaN, sometimes not
- Sometimes the loss will become NaN, sometimes not
- Sometimes some of the intermediate values in the network will be reported to be mathematically impossible values (in particular [this value](https://github.com/chrisc36/jax-debug-nan/blob/7ff6f3b499cf6fc4aceaad2098fce52da16bc98d/t5x/layers.py#L125) will sometimes be much greater than 0), and sometimes not

I can't think of an explanation for this other then some kind of memory-management error, so I am hoping someone here can help me figure out what is going on.

I built a repo to reproduce the issue [here](https://github.com/chrisc36/jax-debug-nan),  it runs a simplified Transformer neural network over a batch of inputs. I would have liked to simplify things more, but so far my attempts have resulted in the NaN disappearing. 

Some other notes:

- I have seen this issue both with and without using `donate_argnums` (I originally saw this bug when a large number of arrays from the optimizer state were donated to the training step), although the repo does not donate anything for simplicity.
- The NaN loss appears only in _very_ particular circumstances, it's not just the amount of memory allocated but the particular shapes of the arrays allocated as well
- Sometime the attention weights the softmax are reported as inf, but the inf does not seem to propagate to other values
- Reproducing the issue has been pretty flaky in general. The repo should reproduce it consistently, but I believe a different kind of TPU will likely see different results

### What jax/jaxlib version are you using?

jax 0.2.27, jaxlib 0.1.76, following T5x

### Which accelerator(s) are you using?

TPU v3-8

**Comment by hawkinsp** [comment]
This sounds concerning. However, those are reasonably old versions of `jax` and `jaxlib`, so the first thing I would want to know is "is this fixed with the latest release", namely jax 0.3.21 and jaxlib 0.3.20?

**Comment by chrisc36** [comment]
I miswrote on in the issue, actually I was tested with jax=0.3.17 and jaxlib=0.3.15

I can confirm the issue continues to exist when upgrading to jax=0.3.21 and jaxlib=0.3.20, I have updated the repo to make sure those versions are installed.

**Comment by skye** [comment]
I'm trying to run your repro, but am getting access denied when trying to download the checkpoint with `gsutil -m cp -r gs://chrisc-public/debug-checkpoint .`. Can you check that it's a public bucket?

**Comment by chrisc36** [comment]
Should be fixed now

**Comment by chrisc36** [comment]
Any updates on this? Even some advice on how to work around or avoid tripping this issue would be helpful, currently its making training our models with jax very difficult.

**Comment by skye** [comment]
@berkinilbeyi is digging into your repro, but we don't have any updates yet.

As a workaround, you could try manually setting `keep_unused=False` here: https://github.com/google/jax/blob/280153334bca32a4fae1f75bd6f1abe1a8f9eacd/jax/experimental/pjit.py#L1068

This will make jax prune unused arguments before compiling with XLA. You can try something like this:
```
git clone https://github.com/google/jax
cd jax
git checkout jax-v0.3.21
pip install -e .
# make local change(s)
```
That should install jax such that any changes you make in that checkout will be reflected when you import jax.

If you give this a try, please let us know how it goes! We could at least make this a configurable option so you don't have to manually hack jax.

**Comment by skye** [comment]
Update: Berkin is making good progress narrowing in on the bug! Another workaround you can try in the meantime: set the env var `LIBTPU_INIT_ARGS='--xla_jf_rematerialization_percent_shared_memory_limit=100'`. This fixes the repro you provided. Berkin's still looking into how to fix this so you don't need any workarounds.

**Comment by chrisc36** [comment]
The env vars seems to work! I tried re-training the model I originally saw the NaN appearing it and it seems to be training smoothly. Thanks for digging into this and finding such a convenient way to fix the issue.

**Comment by skye** [comment]
Great, thanks again for the clear bug report and easy repro! I'll leave this open until we have a real fix committed.
